lib/rpc/UniJsonRpcProvider.ts

```diff
@@ -0,0 +1,316 @@
+import Logger from 'bunyan'
+
+export class UniJsonRpcProvider extends StaticJsonRpcProvider {
+  private readonly chainId: ChainId = ChainId.MAINNET
+
+  private readonly providers: SingleJsonRpcProvider[] = []
+
+  // Used to remember the user-specified precedence or provider URLs. 0 means highest
+  private readonly urlPrecedence: Record<string, number> = {}
+
+  // If provided, we will use this weight to decide the probability of choosing
+  // one of the healthy providers.
+  private readonly urlWeight: Record<string, number> = {}
+
+  private lastUsedProvider: SingleJsonRpcProvider | null = null
+
+  private totallyDisableFallback: boolean = false
+
+  private sessionCache: LRUCache<string, SingleJsonRpcProvider> = new LRUCache({ max: 1000 })
+
+  // If true, it's allowed to use a different provider if the preferred provider isn't healthy.
+  private allowProviderAutoSwitch: boolean = true
+
+  private readonly log: Logger
+
+  constructor(
+    chainId: ChainId,
+    singleRpcProviders: SingleJsonRpcProvider[],
+    log: Logger,
+    ranking?: number[],
+    weights?: number[],
+    allowProviderAutoSwitch?: boolean
+  ) {
+    // Dummy super constructor call is needed.
+    super('dummy_url', { chainId, name: 'dummy_network' })
+    this.connection.url
+    this.log = log
+
+    if (isEmpty(singleRpcProviders)) {
+      throw new Error('Empty singlePrcProviders')
+    }
+
+    if (ranking !== undefined && weights !== undefined && weights && ranking.length != weights.length) {
+      throw new Error('urls and weights need to have the same length')
+    }
+
+    this.chainId = chainId
+    this.providers = singleRpcProviders
+    for (let i = 0; i < this.providers.length; i++) {
+      const url = this.providers[i].url
+      if (ranking !== undefined) {
+        this.urlPrecedence[url] = ranking[i]
+      } else {
+        this.urlPrecedence[url] = i
+      }
+      if (weights != undefined) {
+        if (weights[i] <= 0) {
+          throw new Error(`Invalid weight: ${weights[i]}. Weight needs to be a positive number`)
+        }
+        this.urlWeight[url] = weights[i]
+      }
+    }
+
+    if (allowProviderAutoSwitch !== undefined) {
+      this.allowProviderAutoSwitch = allowProviderAutoSwitch
+    }
+  }
+
+  private selectPreferredProvider(sessionId?: string): SingleJsonRpcProvider {
+    if (this.totallyDisableFallback) {
+      const providers = [...this.providers]
+      this.reorderProviders(providers)
+      this.log.debug(`Use provider ${providers[0].url} for chain ${this.chainId.toString()}`)
+      return providers[0]
+    }
+
+    if (sessionId !== undefined && this.sessionCache.has(sessionId)) {
+      const provider = this.sessionCache.get(sessionId)!
+      if (provider.isHealthy()) {
+        return provider
+      } else if (!this.allowProviderAutoSwitch) {
+        throw new Error(
+          `Forced to use the same provider during the session but the provider (${provider.providerName}) is unhealthy`
+        )
```

Comment:
I think I will log an error instead of throwing here. What if our heuristic is not good enough, and actually provider can service us?